edm4hep.yaml

```
@@ -357,17 +422,30 @@ datatypes:
       - float                energy [GeV]        // energy of the cluster
       - float                energyError [GeV]   // error on the energy
       - edm4hep::Vector3f    position [mm]       // position of the cluster
-      - std::array<float,6>  positionError   // covariance matrix of the position (6 Parameters)
+      - edm4hep::CovMatrix3f positionError   // covariance matrix of the position
       - float                iTheta          // intrinsic direction of cluster at position  Theta. Not to be confused with direction cluster is seen from IP
       - float                phi             // intrinsic direction of cluster at position - Phi. Not to be confused with direction cluster is seen from IP
-      - edm4hep::Vector3f    directionError [mm**2]  // covariance matrix of the direction (3 Parameters)
+      - edm4hep::Vector3f    directionError [mm**2]  // covariance matrix of the direction
     VectorMembers:
       - float   shapeParameters      // shape parameters. This should be accompanied by a descriptive list of names in the shapeParameterNames collection level metadata, as a vector of strings with the same ordering
       - float   subdetectorEnergies  // energy observed in a particular subdetector
     OneToManyRelations:
       - edm4hep::Cluster        clusters     // clusters that have been combined to this cluster
       - edm4hep::CalorimeterHit hits         // hits that have been combined to this cluster
       - edm4hep::ParticleID     particleIDs  // particle IDs (sorted by their likelihood)
+    ExtraCode:
+      includes: "#include <edm4hep/Constants.h>"
+      declaration: "
+      /// Get the position error value for the two passed dimensions\n
+      float getPositionError(edm4hep::Cartesian dimI, edm4hep::Cartesian dimJ) const { return getPositionError().getValue(dimI, dimJ); }\n
+      "
+    MutableExtraCode:
+      includes: "#include <edm4hep/Constants.h>"
+      declaration: "
+      /// Set the position error value for the two passed dimensions\n
+      void setPositionError(float value, edm4hep::Cartesian dimI, edm4hep::Cartesian dimJ) { return getPositionError().setValue(value, dimI, dimJ); }\n
+      void setPositionError(std::array<float, 6> values) { getPositionError() = values; }
```

```suggestion
      void setPositionError(std::array<float, 6> values) { positionError = values; }
```

Doesn't getPositionalError() return `const std::array&` ? Similar in almost all other setters introduced in this PR.

---

It will return a const-correct reference to a `std::array`. This has been introduced with https://github.com/AIDASoft/podio/pull/553

---

OK. [EDM4hep Doxygen](https://edm4hep.web.cern.ch/classedm4hep_1_1_reconstructed_particle.html#a1f6810a43c73575dba7abc54e780db79) hasn't caught up yet I think 

Other thing
```suggestion
      void setPositionError(const std::array<float, 6>& values) { getPositionError() = values; }
```
as far as I understand the assignment operator sets item by item, so this would otherwise make a copy of the input array.
https://en.cppreference.com/w/cpp/container/array
> overwrites every element of the array with the corresponding element of another array

---

> OK. [EDM4hep Doxygen](https://edm4hep.web.cern.ch/classedm4hep_1_1_reconstructed_particle.html#a1f6810a43c73575dba7abc54e780db79) hasn't caught up yet I think

It has; However only the [Mutable types](https://edm4hep.web.cern.ch/classedm4hep_1_1_mutable_reconstructed_particle.html#a7009b2d2db2fe114ab181fe8edea7fc7) get the mutable reference (and the corresponding `const` overload).

> as far as I understand the assignment operator sets item by item, so this would otherwise make a copy of the input array.

Isn't the effect the same in the end? What we want is to have the values that we pass in to be set inside our data. Whether that is done by one copy or by several individual assignments shouldn't matter too much?

---

At least with `-O2` it doesn't seem to make a big difference: [compiler explorer](https://godbolt.org/z/GfWqrKh14). IIUC this both boils down to effectively copying over numbers in chunks of whatever vector register size is available on the underlying hardware.

I also quickly checked with `-O0` as well and the biggest difference seems to be the initial handling of `const&` vs passing by value.

---

So is there any argument against using `const&` ?

---

No, I don't think so.